cmd/microcloud/session.go

```diff
@@ -0,0 +1,166 @@
+package main
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"time"
+
+	"github.com/canonical/microcloud/microcloud/api/types"
+	cloudClient "github.com/canonical/microcloud/microcloud/client"
+	"github.com/canonical/microcloud/microcloud/mdns"
+	"github.com/canonical/microcloud/microcloud/service"
+)
+
+type SessionRole string
+
+const (
+	SessionInitiating SessionRole = "initiating"
+	SessionJoining    SessionRole = "joining"
+)
+
+type SessionFunc func(gw *cloudClient.WebsocketGateway) error
+
+func (c *initConfig) runSession(ctx context.Context, s *service.Handler, role SessionRole, timeout time.Duration, f SessionFunc) error {
+	cloud := s.Services[types.MicroCloud].(*service.CloudService)
+	conn, err := cloud.StartSession(ctx, string(role), timeout)
+	if err != nil {
+		return err
+	}
+
+	defer conn.Close()
+
+	return f(cloudClient.NewWebsocketGateway(ctx, conn))
+}
+
+func (c *initConfig) initiatingSession(gw *cloudClient.WebsocketGateway, sh *service.Handler, services []types.ServiceType, passphrase string, expectedSystems []string) error {
+	session := types.Session{
+		Address:    c.address,
+		Interface:  c.lookupIface.Name,
+		Services:   services,
+		Passphrase: passphrase,
+	}
+
+	err := gw.Write(session)
+	if err != nil {
+		return fmt.Errorf("Failed to send session start: %w", err)
+	}
+
+	err = gw.ReceiveCombined(&session)
+	if err != nil {
+		return fmt.Errorf("Failed to read session reply: %w", err)
+	}
+
+	if !c.autoSetup {
+		fmt.Printf("Enter the passphrase on systems that you want to join the cluster:\n\n %s\n\n", session.Passphrase)
+		fmt.Println("Waiting for systems to send their intent to join ...")
+	}
+
+	confirmedIntents, err := c.askJoinIntents(gw, expectedSystems)
+	if err != nil {
+		return err
+	}
+
+	if !c.autoSetup {
+		cloud := sh.Services[types.MicroCloud].(*service.CloudService)
+		cert, err := cloud.Certificate()
+		if err != nil {
+			return err
+		}
+
+		fingerprint, err := c.shortFingerprint(cert.Fingerprint())
+		if err != nil {
+			return fmt.Errorf("Failed to shorten fingerprint: %w", err)
+		}
+
+		fmt.Printf(" Using fingerprint %q to confirm the join intents\n", fingerprint)
```

Comment:
How about changing this to `Please verify this fingerprint is displayed on all joining systems`. "join intent" should really be an internal concept.

Reply:
Also, can this be moved up where we first display the passphrase? It would be useful to know the fingerprint of the initiator beforehand so that it can be verified on the joiner right after entering the passphrase.

Reply:
Fixed. The used fingerprint is now displayed right after the chosen address for MicroCloud.